pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -332,30 +132,80 @@ func (t *Tracer) install() error {
 	return nil
 }
 
-func (t *Tracer) run(ctx context.Context, logger logger.Logger, dnsTimeout time.Duration) error {
-	spec, err := loadDns()
+func (t *Tracer) parseDNSPacket(rawSample []byte, netns uint64) (*types.Event, error) {
+	// The sample received is a concatenation of the dnsEventT structure and the packet bytes.
+	bpfEvent := (*dnsEventT)(unsafe.Pointer(&rawSample[0]))
+	packetBytes := rawSample[unsafe.Sizeof(*bpfEvent):]
+
+	// The DNS layer starts after the ethernet, IP and UDP headers.
+	DNS_OFF := 14 + 20 + 8
+	dnsLayer := layers.DNS{}
+	err := dnsLayer.DecodeFromBytes(packetBytes[DNS_OFF:], gopacket.NilDecodeFeedback)
 	if err != nil {
-		return fmt.Errorf("loading asset: %w", err)
+		return nil, fmt.Errorf("decoding dns layer: %w", err)
 	}
 
-	parseDNSEvent := func(rawSample []byte, netns uint64) (*types.Event, error) {
-		bpfEvent := (*dnsEventTAbbrev)(unsafe.Pointer(&rawSample[0]))
-		// 4 (padding) + (size of event without addresses) + (address count * 16)
-		expected := 4 + int(unsafe.Sizeof(*bpfEvent)) + int(bpfEvent.Anaddrcount)*16
-		if len(rawSample) != expected {
-			return nil, fmt.Errorf("invalid sample size: received: %d vs expected: %d",
-				len(rawSample), expected)
-		}
+	ipversion := gadgets.IPVerFromAF(bpfEvent.Af)
 
-		event, err := bpfEventToDNSEvent(bpfEvent, rawSample[unsafe.Offsetof(dnsEventT{}.Anaddr):], netns)
-		if err != nil {
-			return nil, err
+	event := &types.Event{
+		Event: eventtypes.Event{
+			Type:      eventtypes.NORMAL,
+			Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+		},
+		WithNetNsID:   eventtypes.WithNetNsID{NetNsID: netns},
+		WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MountNsId},
+		Pid:           bpfEvent.Pid,
+		Tid:           bpfEvent.Tid,
+		Uid:           bpfEvent.Uid,
+		Gid:           bpfEvent.Gid,
+		Comm:          gadgets.FromCString(bpfEvent.Task[:]),
+		PktType:       pktTypeToString(bpfEvent.PktType),
+
+		SrcIP:    gadgets.IPStringFromBytes(bpfEvent.SaddrV6, ipversion),
+		DstIP:    gadgets.IPStringFromBytes(bpfEvent.DaddrV6, ipversion),
+		SrcPort:  bpfEvent.Sport,
+		DstPort:  bpfEvent.Dport,
+		Protocol: gadgets.ProtoString(int(bpfEvent.Proto)),
+
+		ID:         fmt.Sprintf("%.4x", dnsLayer.ID),
+		NumAnswers: int(dnsLayer.ANCount),
+	}
+
+	if dnsLayer.QR {
+		event.Qr = types.DNSPktTypeResponse
+		event.Nameserver = event.SrcIP
+
+		event.Latency = time.Duration(bpfEvent.LatencyNs)
+		event.Rcode = rcodeToString(uint8(dnsLayer.ResponseCode))
+	} else {
+		event.Qr = types.DNSPktTypeQuery
+		event.Nameserver = event.DstIP
+	}
+
+	if len(dnsLayer.Questions) > 0 {
```

Comment:
nit: But probably `len(dnsLayer.Questions)` is more readable:

```suggestion
	if dnsLayer.QDCount > 0 {
```